pandas/io/formats/xml.py

```diff
@@ -0,0 +1,763 @@
+        for root element, conditionally with optional namespaces and
+        prefix.
+        """
+
+        nmsp_dict = {}
+        if self.namespaces and self.prefix is None:
+            nmsp_dict = {"xmlns": n for p, n in self.namespaces.items() if p != ""}
+
+        if self.namespaces and self.prefix:
+            nmsp_dict = {"xmlns": n for p, n in self.namespaces.items() if p == ""}
+
+        return nmsp_dict
+
+    def build_attribs(self) -> None:
+        """
+        Create attributes of row.
+
+        This method adds attributes using attr_cols to row element and
+        works with tuples for multindex or hierarchical columns.
+        """
+
+        for col in self.attr_cols:
+            flat_col = col
+            if isinstance(col, tuple):
+                flat_col = (
+                    "".join(str(c) for c in col).strip()
+                    if "" in col
+                    else "_".join(str(c) for c in col).strip()
+                )
+
+            attr_name = f"{self.prefix_uri}{flat_col}"
+            try:
+                if self.d[col] is not None:
+                    self.elem_row.attrib[attr_name] = str(self.d[col])
+            except KeyError:
+                raise KeyError(f"no valid column, {col}")
+
+    def build_elems(self) -> None:
+        """
+        Create child elements of row.
+
+        This method adds child elements using elem_cols to row element and
+        works with tuples for multindex or hierarchical columns.
+        """
+
+        from xml.etree.ElementTree import SubElement
+
+        for col in self.elem_cols:
+            flat_col = col
+            if isinstance(col, tuple):
+                flat_col = (
+                    "".join(str(c) for c in col).strip()
+                    if "" in col
+                    else "_".join(str(c) for c in col).strip()
+                )
+
+            elem_name = f"{self.prefix_uri}{flat_col}"
+            try:
+                val = None if self.d[col] in [None, ""] else str(self.d[col])
+                SubElement(self.elem_row, elem_name).text = val
+            except KeyError:
+                raise KeyError(f"no valid column, {col}")
+
+    def prettify_tree(self) -> bytes:
+        """
+        Output tree for pretty print format.
+
+        This method will pretty print xml with line breaks and indentation.
+        """
+
+        from xml.dom.minidom import parseString
+
+        dom = parseString(self.out_xml)
+
+        return dom.toprettyxml(indent="  ", encoding=self.encoding)
+
+    def remove_declaration(self) -> None:
+        """
+        Remove xml declaration.
+
+        This method will remove xml declaration of working tree. Currently,
+        pretty_print is not supported in etree.
+        """
+
+        return self.out_xml.split(b"?>")[-1].strip()
+
+    def write_output(self) -> Optional[str]:
+        xml_doc = self.build_tree()
+
+        try:
+            if self.io:
+                with open(self.io, "wb") as f:
```

Comment:
if you want to support compression/fsspec/buffers and so on:

```py
with get_handle(self.io, mode="wb", is_text=False, storage_options=..., compression=...) as handles:
    handles.handle.write(xml_doc)
```

Reply:
I am not strong on these features. I attempted compression in a draft version of `read_xml` looking at `_json.py` but implementation would require workarounds. Can this be for future PR? If not, I would need several days to incorporate.

Reply:
totally fine to defer
just add a check list of todos (top of PR is ok or new issue)